frame/session/src/historical/offchain.rs

```diff
@@ -0,0 +1,272 @@
+
+use super::super::{Module as SessionModule, SessionIndex};
+use super::{IdentificationTuple, ProvingTrie, Trait};
+
+use super::shared;
+use sp_std::prelude::*;
+
+
+/// A set of validators, which was used for a fixed session index.
+pub struct ValidatorSet<T: Trait> {
+	validator_set: Vec<IdentificationTuple<T>>,
+}
+
+impl<T: Trait> ValidatorSet<T> {
+	/// Load the set of validators for a particular session index from the off-chain storage.
+	///
+	/// If none is found or decodable given `prefix` and `session`, it will return `None`.
+	/// Empty validator sets should only ever exist for genesis blocks.
+	pub fn load_from_offchain_db(session_index: SessionIndex) -> Option<Self> {
+		let derived_key = shared::derive_key(shared::PREFIX, session_index);
+		StorageValueRef::persistent(derived_key.as_ref())
+			.get::<Vec<(T::ValidatorId, T::FullIdentification)>>()
+			.flatten()
+			.map(|validator_set| Self { validator_set })
+	}
+
+	/// Access the underlying `ValidatorId` and `FullIdentification` tuples as slice.
+	pub fn as_slice(&self) -> &[(T::ValidatorId, T::FullIdentification)] {
+		self.validator_set.as_slice()
+	}
+
+	/// Convert `self` into a vector and consume `self`.
+	pub fn into_vec(self) -> Vec<(T::ValidatorId, T::FullIdentification)> {
+		self.validator_set
+	}
+
+	/// Attempt to prune anything that is older than `first_to_keep` session index.
+	///
+	/// Due to re-organisation it could be that the `first_to_keep` might be less
+	/// than the stored one, in which case the conservative choice is made to keep records
+	/// up to the one that is the lesser.
+	pub fn prune_older_than(first_to_keep: SessionIndex) {
+		let derived_key = shared::LAST_PRUNE.to_vec();
+		let entry = StorageValueRef::persistent(derived_key.as_ref());
+		match entry.mutate(|current: Option<Option<SessionIndex>>| -> Result<_, ()> {
+			match current {
+				Some(Some(current)) if current < first_to_keep => Ok(first_to_keep),
+				// do not move the cursor, if the new one would be behind ours
+				Some(Some(current)) => Ok(current),
+				None => Ok(first_to_keep),
+				// if the storage contains undecodable data, overwrite with current anyways
+				// which might leak some entries being never purged, but that is acceptable
+				// in this context
+				Some(None) => Ok(first_to_keep),
+			}
+		}) {
+			Ok(Ok(new_value)) => {
+				// on a re-org this is not necessarily true, with the above they might be equal
+				if new_value < first_to_keep {
+					for session_index in new_value..first_to_keep {
+						let derived_key = shared::derive_key(shared::PREFIX, session_index);
+						let _ = StorageValueRef::persistent(derived_key.as_ref()).clear();
+					}
+				}
+			}
+			Ok(Err(_)) => {} // failed to store the value calculated with the given closure
+			Err(_) => {}     // failed to calculate the value to store with the given closure
+		}
+	}
+
+	/// Keep the newest `n` items, and prune all items older than that.
+	pub fn keep_newest(n_to_keep: usize) {
```

Comment:
This doesn't seem to be used either, but I feel it should be.

Reply:
Tracked in #6358